### PR TITLE
smc: calculate investing

### PR DESCRIPTION
**Pull request by lamtt2403**

### CODE REVIEW

File: contracts/CalculateInvesting.sol [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+        uint totalEZC;
+        uint ezcPerDay;
+        uint startTime;
+        bool isLocked;
+        uint numberWithdrawed;
+        uint lastTimeWithdraw;
+    }
+
+    mapping(address => Investment) public investments;
+    address payable public owner;
```
Comment (Zergity) [inline contracts/CalculateInvesting.sol:17]: `owner` don't have to be `payable`

File: contracts/CalculateInvesting.sol [hunk 2]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+contract CalculateInvesting {
+
+    struct Investment {
+        address investor;
+        uint totalEZC;
```
Comment (Zergity) [inline contracts/CalculateInvesting.sol:8]: advanced math and storage optimization: totalEZC, ezcPerDay, startTime, and numberWithdrawed can be redesigned into only 3 uints

File: contracts/CalculateInvesting.sol [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+        owner = payable(msg.sender);
+    }
+
+    modifier onlyOwner() {
+        require(owner == msg.sender, "Only owner");
+        _;
+    }
+
+    function inputInvestment(address _investor, uint _totalEZC, uint _ezcPerDay, uint _startTime) public onlyOwner {
+        investments[_investor] = Investment({
```
Comment (Zergity) [inline contracts/CalculateInvesting.sol:30]: state checks: investment exists, totalEZC > 0, ezcPerDay > 0, etc.

File: contracts/CalculateInvesting.sol [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        });
+    }
+
+    function withdraw() external {
+        Investment storage investment = investments[msg.sender];
+        require(!investment.isLocked, "You are locked!");
+
+        uint numDays = (block.timestamp - investment.startTime) / 60 / 60 / 24;
+        uint limitAmount = investment.totalEZC - investment.numberWithdrawed;
+        uint currentAmount = investment.ezcPerDay * numDays;
```
Comment (Zergity) [inline contracts/CalculateInvesting.sol:47]: wrong calculation, this allows user to withdraw all totalEZC amount without waiting for vesting time

**Review by Zergity — CHANGES_REQUESTED** [review]
1. critical calculation mistake: `limitAmount`
2. missing feature: cliff time
3. missing events
4. missing js feature test for time wrapping test
5. naming semantic: Vesting, not investment (they have total different meaning)

File: contracts/CalculateInvesting.sol [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+            numberWithdrawed: 0,
+            lastTimeWithdraw: 0
+        });
+    }
+
+    function withdraw() external {
+        Investment storage investment = investments[msg.sender];
+        require(!investment.isLocked, "You are locked!");
+
+        uint numDays = (block.timestamp - investment.startTime) / 60 / 60 / 24;
```
[the remark this answers is not in the document]
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol:45]: If you use seconds to calculate then you will have to convert ezcPerDay to ezcPerSecond. Is there any other way?

**Comment by lamtt2403** [comment]
> 1. critical calculation mistake: `limitAmount`
> 2. missing feature: cliff time
> 3. missing events
> 4. missing js feature test for time wrapping test
> 5. naming semantic: Vesting, not investment (they have total different meaning)

**Comment by lamtt2403** [comment]
I just created a new `[pull request](https://github.com/lamtt2403/EZC/pull/2/files)` to update your comments. However, I have not completed the memory optimization (totalEZC, ezcPerDay, startTime, and numberWithdrawed can be redesigned into only 3 uints) yet, I will update later

**Comment by Zergity** [comment]
> I just created a new [pull request](https://github.com/lamtt2403/EZC/pull/2/files) to update your comments. However, I have not completed the memory optimization (totalEZC, ezcPerDay, startTime, and numberWithdrawed can be redesigned into only 3 uints) yet, I will update later

don't do this next time, when you have comment on a PR, you fix it (or discuss it) and push more commit on the PR branch to continue the review

**Comment by lamtt2403** [comment]
> > I just created a new [pull request](https://github.com/lamtt2403/EZC/pull/2/files) to update your comments. However, I have not completed the memory optimization (totalEZC, ezcPerDay, startTime, and numberWithdrawed can be redesigned into only 3 uints) yet, I will update later
> 
> don't do this next time, when you have comments on a PR, you fix it (or discuss it) and push more commits on the PR branch to continue the review

Ok, I got it

File: contracts/CalculateInvesting.sol [hunk 6]
```diff
@@ -0,0 +1,63 @@
+// SPDX-License-Identifier: MIT
+pragma solidity >=0.8.0 <0.9.0;
+
+contract CalculateInvesting {
+
+    struct Investment {
+        address investor;
```
[the remark this answers is not in the document]
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol]: https://github.com/lamtt2403/EZC/commit/1a9c441525ccb40a9e981b623a6f976342c6b378

File: contracts/CalculateInvesting.sol [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+            currentAmount = limitAmount;
+        }
+
+        require(currentAmount != 0, "Amount is not 0");
+        payable(msg.sender).transfer(currentAmount);
+        investment.numberWithdrawed += currentAmount;
+        investment.lastTimeWithdraw = block.timestamp;
+    }
+
+    function changeLock(address _investor) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol]: https://github.com/lamtt2403/EZC/commit/2ae58bf2a6419caa0ce3a26c5ecfe7a4f2ad0e7d

File: contracts/CalculateInvesting.sol [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+            lastTimeWithdraw: 0
+        });
+    }
+
+    function withdraw() external {
+        Investment storage investment = investments[msg.sender];
+        require(!investment.isLocked, "You are locked!");
+
+        uint numDays = (block.timestamp - investment.startTime) / 60 / 60 / 24;
+        uint limitAmount = investment.totalEZC - investment.numberWithdrawed;
```
[the remark this answers is not in the document]
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol]: update: calculate using SafeMath [here](https://github.com/lamtt2403/EZC/commit/2fbf2b1d391ad082be73f13a6955a9a5c7b8120a)

File: contracts/CalculateInvesting.sol (hunk 1 above)
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol]: remove: declared payable [here](https://github.com/lamtt2403/EZC/commit/fb5763f146b340d307fb11249cb1003cc5a6fef9)

File: contracts/CalculateInvesting.sol (hunk 3 above)
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol:39]: state check: https://github.com/lamtt2403/EZC/commit/b3c6ed26af3fae364c48cfc5e43c052bf5e469e3

File: contracts/CalculateInvesting.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (lamtt2403) [reply contracts/CalculateInvesting.sol]: According to the first function, the input is EZC vesting per day. so i will calculate by number days

**Comment by lamtt2403** [comment]
> 1. critical calculation mistake: `limitAmount`
> 2. missing feature: cliff time
> 3. missing events
> 4. missing js feature test for time wrapping test
> 5. naming semantic: Vesting, not investment (they have total different meaning)

cliff time is not mentioned in the requirement

**Pull request closed without merging**